### PR TITLE
Compiler: demonstrate bug from 1354

### PR DESCRIPTION
**Pull request by hhugo**
https://github.com/ocsigen/js_of_ocaml/issues/1354

### CODE REVIEW

File: compiler/tests-compiler/util/util.ml [hunk 1]
```diff
@@ -444,12 +444,12 @@ let compile_and_run_bytecode ?unix s =
...
 
-let compile_and_run ?flags ?unix s =
+let compile_and_run ?debug ?flags ?unix s =
   with_temp_dir ~f:(fun () ->
       s
       |> Filetype.ocaml_text_of_string
       |> Filetype.write_ocaml ~name:"test.ml"
-      |> compile_ocaml_to_bc ?unix
+      |> compile_ocaml_to_bc ?debug ?unix
       |> compile_bc_to_javascript ?flags
```
Comment (vouillon) [inline compiler/tests-compiler/util/util.ml:453]:
```suggestion
      |> compile_bc_to_javascript ?flags ~sourcemap:debug
```

File: compiler/tests-compiler/gh1354.ml [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        var
+         global_data=runtime.caml_get_global_data(),
+         Stdlib=global_data.Stdlib,
+         Stdlib_Format=global_data.Stdlib__Format,
+         _c_=[0,[4,0,0,0,[17,4,0]],caml_string_of_jsbytes("%d@.")],
+         _b_=[0,[4,0,0,0,[17,4,0]],caml_string_of_jsbytes("%d@.")],
+         _a_=0;
+        try
+         {try
+           {var _d_=_a_ + 1 | 0;0;_a_ = _d_;_a_ = _d_;throw Stdlib[3]}
```
Comment (hhugo) [inline compiler/tests-compiler/gh1354.ml:181]: duplicated `assign` need to be fixed  `_a_ = _d_;_a_ = _d_`

File: compiler/tests-compiler/gh1354.ml [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+let%expect_test _ =
+  let prog =
+    {|
+let x = ref 0 in
+let p x = Format.eprintf "%d@." x in
+try
+  incr x;
+  raise Exit
+with Exit ->
+  p !x|}
```
Comment (vouillon) [inline compiler/tests-compiler/gh1354.ml:31]:
```suggestion
  p (!x + 0)|}
```

File: compiler/lib/flow.ml [hunk 4]
```diff
@@ -237,6 +238,7 @@ let program_escape defs known_origins { blocks; _ } =
       List.iter block.body ~f:(fun i ->
           match i with
           | Let (x, e) -> expr_escape st x e
+          | Assign (x, _) -> possibly_mutable.(Var.idx x) <- true
```
Comment (vouillon) [inline compiler/lib/flow.ml:241]: We want `maybe_unknown` to be true (that is, we do not know all possible values of `x`), not `possibly_mutable` to be true (the fields of the values of `x` may be modified).

File: compiler/lib/freevars.ml [hunk 5]
```diff
@@ -53,6 +53,7 @@ let iter_instr_free_vars f i =
       f x;
       f y;
       f z
+  | Assign (_, _) -> ()
```
Comment (vouillon) [inline compiler/lib/freevars.ml:56]:
```suggestion
  | Assign (x, y) -> f x; f y
```

File: compiler/lib/pure_fun.ml [hunk 6]
```diff
@@ -34,6 +34,7 @@ let pure_expr pure_funs e =
 let pure_instr pure_funs i =
   match i with
   | Let (_, e) -> pure_expr pure_funs e
+  | Assign _ -> false
```
Comment (vouillon) [inline compiler/lib/pure_fun.ml:37]: The side-effect cannot be observed from outside the function
```suggestion
  | Assign _ -> true
```

**Comment by vouillon** [comment]
This outputs 1...
```ocaml
let x = ref 0 in
let y = !x in
try
  incr x;
  raise Exit
with Exit ->
  Format.eprintf "%d@." y
```

File: compiler/lib/flow.ml (hunk 4 above)
Comment (vouillon) [reply compiler/lib/flow.ml:241]: Or if we have `Assign (x, y)`, `y` should be a known origin of `x`

**Comment by vouillon** [comment]
The intuition is that `Assign (x, y)` should be like passing an argument `y` to a parameter `x`.

File: compiler/lib/freevars.ml (hunk 5 above)
Comment (vouillon) [reply compiler/lib/freevars.ml]: Actually only `y` is free. `x` is a bound variable (of the exception handler block).

File: compiler/lib/code.ml [hunk 7]
```diff
@@ -317,6 +317,7 @@ type expr =
 
 type instr =
   | Let of Var.t * expr
+  | Assign of Var.t * Var.t
```
[the remark this answers is not in the document]
Comment (vouillon) [reply compiler/lib/code.ml:320]: Probably not worth it, though...

File: compiler/tests-compiler/gh1354.ml (hunk 2 above)
Comment (vouillon) [reply compiler/tests-compiler/gh1354.ml]: This is so rare that it is probably not worth it...

File: compiler/lib/generate.ml [hunk 8]
```diff
@@ -1508,6 +1514,13 @@ and compile_block_no_loop st queue (pc : Addr.t) loop_stack frontier interm =
       let try_catch_frontier = Addr.Set.union new_frontier handler_frontier_cont in
       if debug () then Format.eprintf "@[<hv 2>try {@;";
       if Interm.mem pc1 handler_interm then decr_preds st pc1;
+      (* Parameters of the exception handler are assigned here, not in
+         handler body. Indeed, they correspond to variables that are
+         going to be modified by an instruction Assign. *)
+      let _, assign_handler_variables =
```
Comment (vouillon) [inline compiler/lib/generate.ml:1520]: Alternatively, we could rewrite the code, adding an intermediate branch before the `Pushtrap` than binds the exception handler's variable. Maybe as soon as in `Parse_bytecode`? Then, we do not have to do anything special here.

I'll have to do this transformation to handle effect handlers anyway.

File: compiler/lib/js_assign.ml [hunk 9]
```diff
@@ -316,6 +316,11 @@ let program' (module Strategy : Strategy) p =
   let color = function
     | V v ->
         let name = names.(Var.idx v) in
+        let name =
+          match name with
+          | "" -> "escape"
+          | n -> n
+        in
```
Comment (hhugo) [inline compiler/lib/js_assign.ml:323]: This diff was only for debugging

**Comment by hhugo** [comment]
I recreated the branch by mistake, will reopen an PR.

**Pull request closed without merging**